services/api/lib/bulk.js

```diff
@@ -152,21 +152,39 @@ exports.register = function(server, options, done) {
   // 0th index and the value on that object keyed with 'id'
   var reachRegex = /^\$(\d+)\.(.*)$/;
 
+  var pipelineKeys = [
+    'id',
+    'projectId',
+    'pageId'
+  ];
+
   // check if a key on the action object should be resolved to a value
   // returned by a previous action in the transaction
-  function generateForEachCallback(processResult, action, actionIndex, txResults) {
-    return function everyCallback(key) {
-      // if the key's value isn't a reach string, return valid
-      if ( !reachRegex.test(action.data[key]) ) {
-        return processResult;
+  function reachForData(action, actionIndex, txResults) {
+    var key;
+
+    for (var i = 0; i < pipelineKeys.length; i++) {
+      if (action.data[pipelineKeys[i]]) {
+        key = pipelineKeys[i];
+        break;
       }
+    }
 
-      action.data[key] = action.data[key].replace(
-        reachRegex,
-        getReplaceFunc(processResult, key, actionIndex, txResults)
-      );
-      return processResult;
-    };
+    if (!key) {
+      return true;
+    }
+
+    // if the key's value isn't a reach string, return valid
+    if ( !reachRegex.test(action.data[key]) ) {
+      return true;
+    }
+
+    var replaceResult = {};
```

Comment:
if we are not refactoring this out into a separate function i think all variable declarations should be at the top of the file. not sure what everyone else's feelings are on that; my reasoning: hoisting.